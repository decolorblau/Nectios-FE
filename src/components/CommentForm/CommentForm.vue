<template>
  <v-row justify="center">
    <v-col cols="12">
      <v-card ref="onSubmit" class="pa-2 product-card">
        <v-card-title class="product mt-4 mb-0 pa-0">
          <h2 class="product__title">YOUR OPINION IS IMPORTANT</h2>
        </v-card-title>
        <v-card-text class="pb-0">
          <v-textarea
            ref="description"
            rows="3"
            row-height="18"
            v-model="description"
            maxlength="400"
            :rules="[() => !!description || 'This field is required']"
            label="Write a review"
            counter="400"
            color="#207eeb"
            single-line
            required
            placeholder="Campbell's condensed is a tomato Soup "
            class="ma-0 pa-0"
          ></v-textarea>
        </v-card-text>
        <v-card-actions class="pt-0">
          <v-spacer></v-spacer>
          <v-btn
            variant="outlined"
            class="mt-0 mb-1 mr-2"
            :disabled="description === ''"
            color="#207eeb"
            text
            @click="onSubmit"
          >
            SHARE OPINION
          </v-btn>
        </v-card-actions>
      </v-card>
      <div class="errorMessage">
        <v-alert v-if="wrongProduct" variant="outlined" density="compact" color="#207eeb">
          ! Wrong Comment
        </v-alert>
      </div>
    </v-col>
  </v-row>
</template>

<script>
import { defineComponent } from "vue";
import { mapActions, mapState } from "vuex";

export default defineComponent({
  name: "commentForm",
  data() {
    return {
      description: "",
      isDisabled: true,
      wrongProduct: false,
    };
  },
  computed: {
    ...mapState(["currentProduct", "products"]),
  },
  methods: {
    ...mapActions(["addComment", "getProductComments", "getUserComments", "getProducts"]),
    async onSubmit() {
      try {
        this.wrongProduct = false;
        if (this.description !== "") {
          const { clientKey } = JSON.parse(
            // eslint-disable-next-line comma-dangle
            localStorage.getItem("clientKey") || ""
          );
          const { currentProduct } = JSON.parse(
            // eslint-disable-next-line comma-dangle
            localStorage.getItem("currentProduct") || ""
          );

          const commentData = {
            description: this.description,
            clientKey,
            productKey: currentProduct.key,
          };
          const status = await this.addComment(commentData);
          if (status === 200) {
            this.description = "";
            const productKey = this.currentProduct.product.key;
            this.getProductComments(productKey);
            localStorage.removeItem("phone");
            localStorage.removeItem("userReviews");
            this.getProducts();
            this.getUserComments(this.products);
          } else {
            this.wrongProduct = true;
          }
        }
      } catch {
        this.wrongProduct = true;
      }
    },
  },
});
</script>

<style lang="scss" scoped>
.product {
  text-align: center;
  align-items: center;
  justify-content: center;
  &__title {
    font-size: 16px;
    font-family: "Roboto, Sans-serif";
    font-weight: 400;
  }
}
.product-card {
  box-shadow: 0 0 10px rgba(32, 126, 235, 0.2);
}
.errorMessage {
  margin-top: 10px;
  height: 60px;
  display: flex;
  align-items: center;
  justify-content: center;
}
.blue {
  color: #207eeb;
}
.world {
  font-weight: 600;
}
</style>
