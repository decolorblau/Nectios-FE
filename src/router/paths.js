const paths = {
  home: "/",
  detail: "/products/:id",
  login: "/login",
  newProduct: "/new",
  user: "/user",
  notFound: "/:pathMatch(.*)*",
};

export default paths;
