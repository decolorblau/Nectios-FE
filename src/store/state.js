const state = {
  products: [],
  comments: [],
  user: {
    isAuthenticated: false,
    clientKey: "",
    user: {
      email: "",
      name: "",
      profilePhoto: "",
      role: "",
      surname: "",
    },
  },
  currentProduct: {
    product: {
      id: "",
      client_id: "",
      description: "",
      key: "",
      name: "",
      price: "",
      product_photo: "",
    },
    comments: [],
  },
};

export default state;
