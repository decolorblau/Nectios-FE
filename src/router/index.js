import { createRouter, createWebHistory } from "vue-router";
import Home from "../views/Home.vue";
import Login from "../views/LoginView.vue";
import NotFound from "../views/NotFoundView.vue";
import paths from "./paths";
import User from "../views/UserView.vue";
import Detail from "../views/DetailView.vue";

// eslint-disable-next-line object-curly-newline
const { home, notFound, login, user, detail } = paths;

const routes = [
  {
    path: home,
    name: "Home",
    component: Home,
  },
  {
    path: login,
    name: "Login",
    component: Login,
  },
  {
    path: user,
    name: "User",
    component: User,
  },
  {
    path: detail,
    name: "Detail",
    component: Detail,
  },
  {
    path: notFound,
    name: "NotFound",
    component: NotFound,
  },
];

const router = createRouter({
  history: createWebHistory(process.env.BASE_URL),
  routes,
});

export default router;
