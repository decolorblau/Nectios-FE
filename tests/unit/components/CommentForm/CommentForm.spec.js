import CommentForm from "@/components/CommentForm/CommentForm.vue";
import { shallowMount, mount } from "@vue/test-utils";
import { createStore } from "vuex";
import router from "@/router";
import state from "../../mockState";

describe("Given a CommentForm component", () => {
  describe("When is rendered", () => {
    test("Then it matches de snapshot", () => {
      const wrapper = shallowMount(CommentForm);
      expect(wrapper.html()).toMatchSnapshot();
    });
    test("Then it should render h2 tag with YOUR OPINION IS IMPORTANT text", () => {
      const wrapper = shallowMount(CommentForm);
      const h2Value = wrapper.find("h2").text();
      expect(h2Value).toBe("YOUR OPINION IS IMPORTANT");
    });
    test("Then it should render Description input", () => {
      const wrapper = shallowMount(CommentForm);
      const descriptionInput = wrapper.find("v-textarea");
      const labelDescription = descriptionInput.attributes("label");
      expect(labelDescription).toBe("Write a review");
    });
    test("Then it should render Submit button", () => {
      const wrapper = shallowMount(CommentForm);
      const submitButton = wrapper.find("v-btn").text();
      expect(submitButton).toBe("SHARE OPINION");
    });
    describe("When the form is submitted", () => {
      test("Then it should invoke onSubmit", async () => {
        const store = createStore({
          state() {
            return state;
          },
          actions: {
            loginUser: jest.fn(),
            addComment: jest.fn(),
            getProductComments: jest.fn(),
            getUserComments: jest.fn(),
            getProducts: jest.fn(),
          },
        });

        const wrapper = mount(CommentForm, {
          global: {
            plugins: [router, store],
          },
        });

        const onSubmit = jest.fn();
        onSubmit();
        const form = wrapper.get("v-card");
        await form.trigger("submit");
        expect(onSubmit).toHaveBeenCalled();
      });
    });
  });
});
