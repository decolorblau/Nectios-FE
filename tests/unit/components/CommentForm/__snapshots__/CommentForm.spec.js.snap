// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a CommentForm component When is rendered Then it matches de snapshot 1`] = `
<v-row justify="center">
  <v-col cols="12">
    <v-card class="pa-2 product-card">
      <v-card-title class="product mt-4 mb-0 pa-0">
        <h2 class="product__title">YOUR OPINION IS IMPORTANT</h2>
      </v-card-title>
      <v-card-text class="pb-0">
        <v-textarea rows="3" row-height="18" modelvalue="" maxlength="400" rules="() => {
                      /* istanbul ignore next */
                      cov_rcc3bo5h9.f[9]++;
                      cov_rcc3bo5h9.s[49]++;
                      return (
                        /* istanbul ignore next */
                        (cov_rcc3bo5h9.b[5][0]++, !!_ctx.description) ||
                        /* istanbul ignore next */
                        (cov_rcc3bo5h9.b[5][1]++, 'This field is required')
                      );
                    }" label="Write a review" counter="400" color="#207eeb" single-line="" required="" placeholder="Campbell's condensed is a tomato Soup " class="ma-0 pa-0"></v-textarea>
      </v-card-text>
      <v-card-actions class="pt-0">
        <v-spacer></v-spacer>
        <v-btn variant="outlined" class="mt-0 mb-1 mr-2" disabled="true" color="#207eeb" text=""> SHARE OPINION </v-btn>
      </v-card-actions>
    </v-card>
    <div class="errorMessage">
      <!--v-if-->
    </div>
  </v-col>
</v-row>
`;
